src/buffer.cc

```diff
@@ -30,16 +33,104 @@ FormatBuffer* Buffer::AsFormatBuffer() {
   return static_cast<FormatBuffer*>(this);
 }
 
+Result Buffer::CopyBaseFieldsTo(Buffer* buffer) const {
+  if (buffer->width_ != width_)
+    return Result("Buffer::CopyBaseFields() buffers have a different width");
+  if (buffer->height_ != height_)
+    return Result("Buffer::CopyBaseFields() buffers have a different height");
+  if (buffer->size_ != size_)
+    return Result("Buffer::CopyBaseFields() buffers have a different size");
+  buffer->data_.clear();
+  buffer->data_ = data_;
+  buffer->values_.clear();
+  buffer->values_ = values_;
+  return {};
+}
+
+Result Buffer::CopyTo(Buffer* buffer) const {
+  Result result = CopyBaseFieldsTo(buffer);
+  if (!result.IsSuccess())
+    return result;
+
+  if (buffer->mem_ptr_) {
+    // TODO(hevrard) Return error if buffer_to mem_ptr_ is backed by GPU memory
+    std::free(buffer->mem_ptr_);
+  }
+  buffer->mem_ptr_ = std::malloc(GetSizeInBytes());
+  if (!buffer->mem_ptr_)
+    return Result("Buffer::CopyTo() malloc failed");
+  std::memcpy(buffer->mem_ptr_, mem_ptr_, GetSizeInBytes());
+
+  return {};
+}
+
 DataBuffer::DataBuffer() = default;
 
 DataBuffer::DataBuffer(BufferType type) : Buffer(type) {}
 
 DataBuffer::~DataBuffer() = default;
 
+Result DataBuffer::CopyTo(Buffer* buffer) const {
+  Result result = CopyBaseFieldsTo(buffer);
+  if (!result.IsSuccess())
+    return result;
+
+  if (!buffer->IsDataBuffer())
+    return Result(
+        "DataBuffer::CopyTo() destination buffer is not a data buffer");
+
+  auto data_buffer = buffer->AsDataBuffer();
+
+  if (data_buffer->datum_type_.GetType() != datum_type_.GetType())
+    return Result("DataBuffer::CopyTo() buffers have a different datum type");
+  if (data_buffer->GetSizeInBytes() != GetSizeInBytes())
+    return Result("DataBuffer::CopyTo() buffers have a different byte size");
+
+  if (data_buffer->mem_ptr_) {
```

Comment:
ditto

Reply:
Actually, you could add a help in Buffer which does the mem_ptr_ copy. Just send in the size from the overridden CopyTo method. Removes the duplicated code.

Reply:
Good point!